Developer Notes! (#750)
svof/svof: Developer_readme.md

> # Source code layout

Reviewer: Could you manually wrap / change formatting of this? Going left and right kinda tough:

https://user-images.githubusercontent.com/110988/146669732-4fc2e7ce-7f52-4575-8f6b-ae6c006ca19f.mp4

Author: Ok I edited it, see if it is okay that way?

3. add new diagnose trigger for it on **svo (aliases, triggers)**.
5. add to herb cures/restoration/smoke/focus/humour/etc tables (whatever that is applicable) on **svo (trigger functions) > Main trigger functions** (the comment lines `-- normal herbs`, `-- normal smokes`. `-- focus`, etc can help you find your way).
7. add to generic cures (passive cures or cures that happen in blackout) (generic_cures_data in **Main trigger functions** on **svo (trigger functions)**)
8. add its definition on svof-serverside integration (sstosvoa on dict).
9. (Optional) add triggers receiving and losing affliction. Note: This should only be necessary for afflictions that are not shown on GMCP for some reason (best example I can think of is Pariah's latency), the system already handles gaining/removing aff through GMCP so no need to add triggers for that on normal circumstances.
10. (Optional) check failure conditions and add them (in case it is an affliction that is not trackeable on gmcp). You can add the logic on the dict entry or on triggers, wherever applicable.
<details>
    <summary>This is an example of how an affliction entry is structured: </summary>
    
```lua
  paralysis = {

>     herb = {

Reviewer: ```suggestion
    herb = { -- balance to cure / do the action with - in this case, cure paralysis with herbs
```

>       isadvisable = function() -- add all the conditions necessary for the system to consider that you have been afflicted by this here

Reviewer: ```suggestion
      isadvisable = function() -- add all the conditions necessary for the system to consider that you can & should cure it with this balance. Some checks common to all actions on this balance aren't repeated here (for example herbs won't check if you have herb balance here, that'd be repetitive)
```

>         return (affs.paralysis) or false -- means if you are paralysed already, it won't double paralysis

Reviewer: ```suggestion
        return (affs.paralysis) or false -- run this action if you've got paralysis, some of these can get pretty complex
```

      end,

      oncompleted = function() -- affliction got cured, signal svo to remove it from the affs list and that herb balance has been lost since you just ate the herb to cure this, add other logic if necessary
        svo.rmaff('paralysis')
        svo.lostbal_herb()
        svo.killaction(svo.dict.checkparalysis.misc)
      end,

      eatcure = {'bloodroot', 'magnesium'}, -- tells svo what kind of cure to make this go away

>       onstart = function() svo.eat(svo.dict.paralysis.herb) end, -- svo started the curing procedure, tell it what it should do (in this case, it will eat the specified cure above)

Reviewer: ```suggestion
      onstart = function() svo.eat(svo.dict.paralysis.herb) end, -- svo started the curing procedure because `isadvisable()` above said it should, tell it what it should do (in this case, it will eat the specified `eatcure`)
```

>       empty = function() empty.eat_bloodroot() end -- for empty cure handling

Reviewer: ```suggestion
      empty = function() empty.eat_bloodroot() end -- what to do when you ate the herb but no affliction got cured
```

    },
    aff = { -- additional affliction behavior logic
      oncompleted = function() -- to be executed upon gaining this aff
        svo.addaffdict(svo.dict.paralysis)
        signals.after_lifevision_processing:unblock(cnrl.checkwarning)
      end,
    },

>     gone = { -- logic to be executed after paralysis is cured, in this case, just remove it from the aff list

Reviewer: ```suggestion
    gone = { -- logic to be executed when we get a line from the game saying the aff is gone/cured, in this case, just remove it from the aff list
```

      oncompleted = function() svo.rmaff('paralysis') end,
    },

>     onremoved = function() svo.affsp.paralysis = nil svo.donext() end -- additional logic to be executed after paralysis has been sucessfully removed/cured, means the system can now do any schedule actions

Reviewer: ```suggestion
    onremoved = function() svo.affsp.paralysis = nil svo.donext() end -- additional logic to be executed after paralysis has been removed from the aff list (so after the gone.oncompleted right above ran), means the system can now do any schedule 'do' actions
```

1. update madness_affs table in empty cures on **svo (setup, misc, empty, funnies, dor) > Empty cure handling**.

## How to add a new defence
1. add it in **svo (actions dictionary)** in the dict table, with the appropriate functions and defup/keepup logic.
    - There are two ways of making svo recognize a defence dict entry: through the basicdef function and through a hardcoded entry. Use basicdef whenever you have a simple defence that can be turned on or off without any special logic required. Alternatively, use a dictionary entry whenever you need custom defence logic. Examples:
    
    basicdef example:
    
    ```lua
    if svo.haveskillset('curses') then
      basicdef('swiftcurse', 'swiftcurse') -- refer to the notes on the svo.basicdef function to see the usage, this basically means 'defencename', 'command'
    end
    ```
    
    dictionary entry with custom defence logic:
    ```lua
      svo.dict.devour = {
        gamename = 'devour', -- ingame name
        physical = {
          balanceful_act = true, def = true, -- means it is a def that takes balance and flags this dict entry as a defence entry

>           isadvisable = function() -- logic that tells svo when to put this defence

Reviewer: ```suggestion
          isadvisable = function() -- logic that tells svo when to put up this defence
```

            return (not defc.dragonform and not defc.devour and ((sys.deffing and defdefup[defs.mode].devour) or (conf.keepup and defkeepup[defs.mode].devour)) and not codepaste.balanceful_defs_codepaste() and not affs.paralysis and not affs.prone and (defc.mouths and defc.tentacles) and bals.anathema) or false
          end,
          -- ^ means svo will only put this defence if the user is not dragon, don't already have devour up, is not in the process of putting up or keeping up a defence at the time, has passed balanceful defs checks, is not paralysed and not prone, has the mouths and tentacles defences already up and has anathema balance available. 

          oncompleted = function() defences.got('devour') end, -- defence is sucessfully up? tell svo you got it.

          action = "unnamable devour", -- defines the action that will put this defence up

>           onstart = function() -- logic to put the defence up, add what needs to be sent and any other applicable logic to it here.
>             send('unnamable devour', conf.commandecho) 
>             end

Reviewer: ```suggestion
          onstart = function() -- logic to put the defence up, add what needs to be sent and any other applicable logic to it here.
            send('unnamable devour', conf.commandecho) 
          end
```

  defs_data:set('deathsight', { type = 'necromancy', -- the type determines from what skillset the defence will be put into, usually you want the type to be the same name as the skillset you are adding it for
    staysindragon = true, -- means that the defence will stay on even after dragonforming
    availableindragon = true, -- means it can be putup while in dragonform. For this specific case, 'deathsight' is also a general defence that can be put up in dragonform so this flag this defence so svo to not unnecessarily drop it after the defence table is repopulated upon dragonforming
    def = "Your mind has been attuned to the realm of Death.", -- the line that appears when you check DEF ingame
    on = {"Your mind is already attuned to the realm of Death.", "You shut your eyes and concentrate on the Soulrealms. A moment later, you feel inextricably linked with realm of Death."}, -- these are the lines that are fired when you put the defence up. 

>     onr = "^A miasma of darkness passes over your eyes and you feel a link to the realm of Death,? form in your mind\\.$", --same as above, except that it uses regex. Use onr anytime you need regex instead of a normal trigger.

Reviewer: Should this be just one `\`?

Author: umm i dunno, I copied it straight from the code itself! Maybe it was a typo?

  defs_data:set('lifevision', { type = 'necromancy',
    on = {"You narrow your eyes and blink rapidly, enhancing your vision to seek out sources of lifeforce in others.", "You already possess enhanced vision."},
    def = "You have enhanced your vision to be able to see traces of lifeforce."})
  defs_data:set('putrefaction', { type = 'necromancy',
    on = {"You concentrate for a moment and your flesh begins to dissolve away, becoming slimy and wet.", "You have already melted your flesh. Why do it again?"},
    def = "You are bathed in the glorious protection of decaying flesh.",
    off = "You concentrate briefly and your flesh is once again solid."})
  defs_data:set('vengeance', { type = 'necromancy',
    staysindragon = true,
    offline_defence = true,
    on = {"You swear to yourself that you will wreak vengeance on your slayer.", "Vengeance already burns within your heart, Necromancer."},
    def = "You have sworn vengeance upon those who would slay you.",
    off = {"You forswear your previous oath for vengeance, sudden forgiveness entering your heart.", "You have sworn vengeance against none, Necromancer."}})
end
```
</details>

Important notes:

> - You can use defr, onr and offr whenever you need to use regex for def, defup and defoff messages respectively. They are not mutually exclusive with normal def, on, off messages so you can use normal and regex triggers on the same entry.

Reviewer: ```suggestion
- You can use `defr`, `onr` and `offr` whenever you need to use regex for def, defup and defoff messages respectively. They are not mutually exclusive with normal `def`, `on`, `off` messages so you can use normal and regex triggers on the same entry.
```

- Available flags and configuration functions are:

>     - availableindragon -> boolean value that flags when a defence is available and can be put up even in dragonform (usually only for defences that also shares a general or dragon equivalent with the same name and behavior, like deathsight.

Reviewer: ```suggestion
    - `availableindragon` -> boolean value that flags when a defence is available and can be put up even in dragonform (usually only for defences that also shares a general or dragon equivalent with the same name and behavior, like deathsight.
```

>     - invisibledef -> boolean value that determines whether the defence is invisible (not recognized as a real defence by achaea). This is useful for certain abilities that are not naturally defences per se, but still useful to keep track of.

Reviewer: ```suggestion
    - `invisibledef` -> boolean value that determines whether the defence is invisible (not recognized as a real defence by achaea). This is useful for certain abilities that are not naturally defences per se, but still useful to keep track of.
```

>     - offline_defence -> boolean value that flags defences that can stay up even after logging out, see the comments in the code above for example

Reviewer: ```suggestion
    - `offline_defence` -> boolean value that flags defences that can stay up even after logging out, see the comments in the code above for example
```

>     - on_enable -> custom function, this is mostly only used for monks and blademasters to assure the form they are in can use said defence. See `retaliationstrike` as an example. Can become useful in case new classes that contain different forms as well are launched.

Reviewer: ```suggestion
    - `on_enable` -> custom function, this is mostly only used for monks and blademasters to assure the form they are in can use said defence. See `retaliationstrike` as an example. Can become useful in case new classes that contain different forms as well are launched.
```

> Can become useful in case new classes that contain different forms as well are launched.

I think we're missing a word here

Author: Ah, yeah I will reword that it got a bit confusing

>     - specialskip -> flags defences that for some reason you want to ignore from defup (meaning svo will skip them instead of trying to put them up before proceeding to the next defences in queue to be put up)

Reviewer: ```suggestion
    - `specialskip` -> flags defences that for some reason you want to ignore from defup in order not to get stuck (meaning svo will skip them instead of trying to put them up before proceeding to the next defences in queue to be put up)
```

>     - stays_in_dragon -> boolean value that determines whether or not the defence stays up after dragonforming

Reviewer: ```suggestion
    - `stays_in_dragon` -> boolean value that determines whether or not the defence stays up after dragonforming
```

>     - stays_on_death -> boolena value that flags this defence as one that stays up even upon death

Reviewer: ```suggestion
    - `stays_on_death` -> boolean value that flags this defence as one that stays up even upon death
```

>     - mana -> optional setting that sets the mana usage for said defence: 'lots' for high mana usag, 'little' for low mana usage. This will put an uppercase 'M' or a lowercase 'm' respectively on vshow defup/keepup for defences that drains mana.

Reviewer: ```suggestion
    - `mana` -> optional setting that sets the mana usage for said defence: 'lots' for high mana usag, 'little' for low mana usage. This will put an uppercase 'M' or a lowercase 'm' respectively on vshow defup/keepup for defences that drains mana.
```
